tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,283 @@
+      << indent << "M=" << params.M << "\n"
+      << indent << "K=" << params.K << "\n"
+      << indent << "N=" << params.N << "\n"
+      << indent << "S=" << params.S << "\n"
+      << indent << "use_different_streams=" << params.use_different_streams
+      << "\n"
+      << "}";
+  return out;
+}
+
+class OverlapTest : public MultiDeviceTest {
+ protected:
+  OverlapTestParams params;
+
+  int64_t num_devices;
+  int64_t my_device_index;
+  at::Tensor ta, tb, tc_unreduced, tc_locally_reduced, tc, tc_expected;
+  // stores the backend
+  c10d::Backend* world_communicator;
+
+  void SetUp() {
+    MultiDeviceTest::SetUp();
+
+    params.parseEnv();
+    num_devices = communicator->size();
+    my_device_index = communicator->deviceId();
+    ASSERT_EQ(params.M % (params.S * num_devices), 0);
+    ASSERT_EQ(params.K % num_devices, 0);
+
+    // Setup the world communicators
+    std::vector<int64_t> devices(num_devices);
+    std::iota(devices.begin(), devices.end(), 0);
+    world_communicator = communicator->getBackendForTeam(
+        devices, /*backend=*/params.backend_type);
+
+    // Define I/O and intermediate Tensor shapes
+    // clang-format off
+        std::vector<int64_t> ta_unsharded_sizes         = {params.S, num_devices, params.M/params.S, params.K/num_devices, 1       };
+        std::vector<int64_t> ta_sizes                   = {params.S, 1          , params.M/params.S, params.K/num_devices, 1       };
+        std::vector<int64_t> tb_unsharded_sizes         = {1       , num_devices, 1                , params.K/num_devices, params.N};
+        std::vector<int64_t> tb_sizes                   = {1       , 1          , 1                , params.K/num_devices, params.N};
+        std::vector<int64_t> tc_unreduced_sizes         = {params.S, 1          , params.M/params.S, params.K/num_devices, params.N};
+        std::vector<int64_t> tc_partially_reduced_sizes = {params.S, 1          , params.M/params.S,                       params.N};
+        std::vector<int64_t> tc_sizes                   = {params.S, 1          , params.M/(params.S*num_devices)        , params.N};
+    // clang-format on
+
+    // Set up input tensors. We create the full unsharded tensors and define the
+    // actual input as the shard corresponding to the current device. Having the
+    // full unsharded input on each rank makes it possible to compute the
+    // expected result locally, hence, this way of doing is convenient for
+    // validating data correctness.
+    at::TensorOptions options =
+        at::TensorOptions().dtype(at::kFloat).device(communicator->device());
+    auto ta_unsharded = at::randn(ta_unsharded_sizes, options);
+    auto tb_unsharded = at::randn(tb_unsharded_sizes, options);
+    ta = at::empty(ta_sizes, options);
+    ta.copy_(ta_unsharded.index(
+        {at::indexing::Slice(),
+         at::indexing::Slice(my_device_index, my_device_index + 1),
+         "..."}));
```

Comment:
How is it different from `ta_unsharded.slice(1, my_device_index, my_device_index + 1)`, which is simpler?

Reply:
let me fix that